resolver/src/lookup_ip.rs

```diff
@@ -81,12 +80,68 @@ where
     names: Vec<Name>,
     strategy: LookupIpStrategy,
     options: DnsRequestOptions,
-    future: Box<Future<Item = Lookup, Error = ResolveError> + Send>,
+    query: Box<Future<Item = Lookup, Error = ResolveError> + Send>,
     hosts: Option<Arc<Hosts>>,
     finally_ip_addr: Option<RData>,
 }
 
-impl<C: DnsHandle<Error = ResolveError> + 'static> LookupIpFuture<C> {
+impl<C: DnsHandle<Error = ResolveError> + 'static> Future for LookupIpFuture<C> {
+    type Item = LookupIp;
+    type Error = ResolveError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            // Try polling the underlying DNS query.
+            let query = self.query.poll();
+
+            // Determine whether or not we will attempt to retry the query.
+            let should_retry = match query {
+                // If the query is NotReady, yield immediately.
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                // If the query returned a successful lookup, we will attempt
+                // to retry if the lookup is empty. Otherwise, we will return
+                // that lookup.
+                Ok(Async::Ready(ref lookup)) => lookup.is_empty(),
+                // If the query failed, we will attempt to retry.
+                Err(_) => true,
+            };
+
+            if should_retry {
+                if let Some(name) = self.names.pop() {
+                    // If there's another name left to try, build a new query
+                    // for that next name and continue looping.
+                    self.query = strategic_lookup(
```

Comment:
My only concern here is that we need to guarantee that this lookup call itself polls for readiness, otherwise we could end up not getting rescheduled in an async(ready) case.

Can you add a comment about that here? I think our tests should catch an error in this situation...

Reply:
Okay, what do you think of https://github.com/bluejekyll/trust-dns/pull/480/commits/199ae2e0347ff5a0db9341400199769b92911959#diff-59fbbc4979234163c73268d87add0179R95? :)

Reply:
Thanks you for your explanation! And the additional comment. That’s perfect.

Also it’s made me appreciate the loop you’ve added here. That’s a very elegant solution to this pattern. I’m going to need to apply that throughout the code! Nice work!